src/tasks/MonoTargetsTasks/ILStrip/ILStrip.cs

```diff
@@ -81,4 +98,177 @@ private bool StripAssembly(ITaskItem assemblyItem)
+            {
+                method_body_uses[rva]++;
+            }
+            else
+            {
+                method_body_uses.Add(rva, 1);
+            }
+        }
+
+        do
+        {
+            int methodToken2Trim = Convert.ToInt32(line, 16);
+            int rva2Trim = token_to_rva[methodToken2Trim];
+            method_body_uses[rva2Trim]--;
+        } while ((line = sr.ReadLine()) != null);
+
+        return method_body_uses;
+    }
+
+    private static void CreateTrimmedAssembly(PEReader peReader, string trimmedAssemblyFilePath, FileStream fs, Dictionary<int, int> method_body_uses)
+    {
+        using (FileStream os = File.Open(trimmedAssemblyFilePath, FileMode.Create))
+        {
+            fs.Position = 0;
+            MemoryStream memStream = new MemoryStream((int)fs.Length);
+            fs.CopyTo(memStream);
+
+            foreach (var kvp in method_body_uses)
+            {
+                int rva = kvp.Key;
+                int count = kvp.Value;
+                if (count == 0)
+                {
+                    int methodSize = ComputeMethodSize(peReader, rva);
+                    int actualLoc = ComputeMethodHash(peReader, rva);
+                    int headerSize = ComputeMethodHeaderSize(memStream, actualLoc);
+                    ZeroOutMethodBody(ref memStream, methodSize, actualLoc, headerSize);
+                }
+            }
+
+            memStream.Position = 0;
+            memStream.CopyTo(os);
+        }
+    }
+
+    private static int ComputeMethodSize(PEReader peReader, int rva)
+    {
+        MethodBodyBlock mb = peReader.GetMethodBody(rva);
+        return mb.Size;
+    }
+
+    private static int ComputeMethodHash(PEReader peReader, int rva)
+    {
+        int sectionIndex = peReader.PEHeaders.GetContainingSectionIndex(rva);
+        int relativeOffset = rva - peReader.PEHeaders.SectionHeaders[sectionIndex].VirtualAddress;
+        return (peReader.PEHeaders.SectionHeaders[sectionIndex].PointerToRawData + relativeOffset);
+    }
+
+    private static int ComputeMethodHeaderSize(MemoryStream memStream, int actualLoc)
+    {
+        memStream.Position = actualLoc;
+        int firstbyte = memStream.ReadByte();
+        int headerFlag = firstbyte & 0b11;
+        return (headerFlag == 2 ? 1 : 4);
+    }
+
+    private static void ZeroOutMethodBody(ref MemoryStream memStream, int methodSize, int actualLoc, int headerSize)
+    {
+        memStream.Position = actualLoc + headerSize;
+
+        byte[] zeroBuffer;
+        zeroBuffer = ArrayPool<byte>.Shared.Rent(methodSize);
+        Array.Clear(zeroBuffer, 0, zeroBuffer.Length);
+        memStream.Write(zeroBuffer, 0, methodSize - headerSize);
+        ArrayPool<byte>.Shared.Return(zeroBuffer);
+    }
+
+    private static void ReplaceAssemblyWithTrimmedOne(string assemblyFilePath, string trimmedAssemblyFilePath)
+    {
+        File.Delete(assemblyFilePath);
+        File.Move(trimmedAssemblyFilePath, assemblyFilePath);
+    }
```

Comment:
This does make me a little nervous. We don't know where that original assembly file is located, and if it's in use, like Aleksey mentioned. Maybe we should emit these into a separate temporary folder, output the list of these updated files from the task, and let the calling targets take care of this. The targets have a better chance for having context about the files.

Reply:
Are MSBuild tasks running sequentially? If so, why would they be in use?

Reply:
The tasks run sequentially for a *project*. But my concern is that we don't know what else is really using the those files. And the task itself doesn't have the context. but calling target might, and that can move the files right after the task is done too. A different use of the same *task* might need to do something else. IOW, the task has no idea in what kinda context it was invoked.